hubblestack/loader.py

```diff
@@ -0,0 +1,1292 @@
+    if 'conf_file' in opts:
+        pre_opts = {}
+        pre_opts.update(salt.config.load_config(
+            opts['conf_file'], 'SALT_MINION_CONFIG',
+            salt.config.DEFAULT_MINION_OPTS['conf_file']
+        ))
+        default_include = pre_opts.get(
+            'default_include', opts['default_include']
+        )
+        include = pre_opts.get('include', [])
+        pre_opts.update(salt.config.include_config(
+            default_include, opts['conf_file'], verbose=False
+        ))
+        pre_opts.update(salt.config.include_config(
+            include, opts['conf_file'], verbose=True
+        ))
+        if 'grains' in pre_opts:
+            opts['grains'] = pre_opts['grains']
+        else:
+            opts['grains'] = {}
+    else:
+        opts['grains'] = {}
+
+    grains_data = {}
+    funcs = grain_funcs(opts, proxy=None)
+    if force_refresh:  # if we refresh, lets reload grain modules
+        funcs.clear()
+    # Run core grains
+    for key in funcs:
+        if not key.startswith('core.'):
+            continue
+        log.trace('Loading %s grain', key)
+        ret = funcs[key]()
+        if not isinstance(ret, dict):
+            continue
+        if grains_deep_merge:
+            salt.utils.dictupdate.update(grains_data, ret)
+        else:
+            grains_data.update(ret)
+
+    # Run the rest of the grains
+    for key in funcs:
+        if key.startswith('core.') or key == '_errors':
+            continue
+        try:
+            # Grains are loaded too early to take advantage of the injected
+            # __proxy__ variable.  Pass an instance of that LazyLoader
+            # here instead to grains functions if the grains functions take
+            # one parameter.  Then the grains can have access to the
+            # proxymodule for retrieving information from the connected
+            # device.
+            log.trace('Loading %s grain', key)
+            parameters = salt.utils.args.get_function_argspec(funcs[key]).args
+            kwargs = {}
+            if 'proxy' in parameters:
+                kwargs['proxy'] = proxy
+            if 'grains' in parameters:
+                kwargs['grains'] = grains_data
+            ret = funcs[key](**kwargs)
+        except Exception:
+            if salt.utils.platform.is_proxy():
+                log.info('The following CRITICAL message may not be an error; the proxy may not be completely established yet.')
+            log.critical(
+                'Failed to load grains defined in grain file %s in '
+                'function %s, error:\n', key, funcs[key],
+                exc_info=True
+            )
+            continue
+        if not isinstance(ret, dict):
+            continue
+        if grains_deep_merge:
+            salt.utils.dictupdate.update(grains_data, ret)
+        else:
+            grains_data.update(ret)
+
+    if opts.get('proxy_merge_grains_in_module', True) and proxy:
```

Comment:
I believe this is also for proxy minions.

Reply:
indeed